04 SFC Recipe List/README.md

```diff
@@ -0,0 +1,693 @@
+     3. Cook pasta according to package directions. Drain. Serve sauce over spaghetti.
+    `,
+    ingredients: [
+      'spaghetti',
+      '1 beef',
+      '1/4 teaspoon garlic powder',
+      '1 onion',
+      '3 tomatoes',
+      '4 mushrooms',
+      'oregano',
+      'salt',
+    ],
+  },
+];
+
+```
+
+### ./src/rest-api/api/recipe/recipe.ts
+
+```javascript
+import { Recipe } from '../../model';
+import { mockRecipes } from './mockData';
+
+export const fetchRecipes = (): Promise<Recipe[]> => {
+  return Promise.resolve(mockRecipes);
+};
+
+```
+
+### ./src/rest-api/api/recipe/index.ts
+
+```javascript
+export * from './recipe';
+
+```
+
+- Create `recipe` viewModel:
+
+### ./src/pages/recipe/list/viewModel.ts
+
+```javascript
+export interface Recipe {
+  id: number;
+  name: string;
+  description: string;
+  ingredients: string[];
+};
+
+```
+
+- Create `mapper` to map from model to viewModel:
+
+### ./src/pages/recipe/list/mappers.ts
+
+```javascript
+import * as model from '../../../rest-api/model';
+import * as vm from './viewModel';
+
+export const mapRecipeListModelToVm = (recipes: model.Recipe[]): vm.Recipe[] => (
+  recipes.map(mapRecipeModelToVm)
+);
+
+const mapRecipeModelToVm = (recipe: model.Recipe): vm.Recipe => ({
+  ...recipe,
+});
+
+```
+
+- Create `recipe list` page container:
+
+### ./src/pages/recipe/list/PageContainer.vue
+
+```javascript
+<template>
+  <recipe-list-page
+    :recipes="recipes"
+  />
+</template>
+
+<script lang="ts">
+import Vue from 'vue';
+import { Recipe } from './viewModel';
+import { mapRecipeListModelToVm } from './mappers';
+import { fetchRecipes } from '../../../rest-api/api/recipe';
+import RecipeListPage from './Page.vue';
+
+export default Vue.extend({
+  name: 'RecipeListPageContainer',
+  components: {
+    RecipeListPage,
+  },
+  data() {
+    return {
+      recipes: [] as Recipe[],
+    };
+  },
+  created() {
+    fetchRecipes()
+      .then(recipes => {
+        this.recipes = mapRecipeListModelToVm(recipes);
+      })
+      .catch(error => console.log(error));
+  },
+});
+</script>
+
+```
+
+- Create `header`:
+
+### ./src/pages/recipe/list/components/Header.vue
+
+```javascript
+<template>
+  <thead>
+    <th>
+      Name
+    </th>
+    <th>
+      Description
+    </th>
+    <th>
+    </th>
+  </thead>
+</template>
+
+<script lang="ts">
+import Vue from 'vue';
+
+export default Vue.extend({
+  name: 'HeaderComponent',
+});
+</script>
+
+```
+
+- To work with CSS Modules, we'll install `webpack-env` typings for use `require`:
+
+```
+npm install @types/webpack-env --save-dev
+```
+
+- Configure CSS modules:
+
+### ./webpack.config.js
+
+```diff
+  ...
+  ...
+  module: {
+    rules: [
+      ...
+      {
+        test: /\.css$/,
+-       use: [
+-         process.env.NODE_ENV !== 'production'
+-           ? 'vue-style-loader'
+-           : MiniCssExtractPlugin.loader,
+-         'css-loader'
+-       ],
++       oneOf: [
++         {
++           resourceQuery: /module/,
```

Comment:
What does `/module/` mean? Styles from `node_modules`?

Reply:
Nvm, thank you for the info link about css modules in `.vue` files. It means we could define components with CSS Modules like: 

```vue
<template>
...
</template>
<script>
...
</script>
<style module> // here place css styles only available for this component
...
</style>
```

About the `oneOf` webpack rule to get a fallback for styles not defined with css modules, do you think that is it necessary?
I mean, if we are using  css modules, we are going to define all components using it, isn't it? 
Except some global styles, but we will import that styles in webpack.config.js like:

```javascript
app: './main.ts',
    vendor: [
      'lc-form-validation',
      'vue',
      'vue-router',
    ],
    appStyles: [
      './styles/main.css',
    ],
    vendorStyles: [
      '../node_modules/bootstrap/dist/css/bootstrap.css',
    ],
```

What do you think?